### PR TITLE
Separate Archive Scan argument related to security from `--force`

### PR DESCRIPTION
**Pull request by Trenly**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ ] Are you working against an Issue?

A change was recently made to make the hash override argument different from the `--force` argument on the basis that it has heavy implications on security and that those security implications should not be as easy to bypass as just adding on a "standard" parameter. This PR does the same thing for the archive malware scan.

This is implemented as a more generic `ignore-security-checks` argument. The reason for this is twofold - 1) The archive scan comes after the hash validation. The security implications of skipping this check (as opposed to the hash check) are lower, and therefore a less-specific argument can be used (just something less generic than `--force`) and 2) This argument will be expansible to other security-related actions. One such example would be if/when CVE data is integrated, a user would be able to use the new argument to confirm they wish to install a package with a known vulnerability.

This is not a breaking change as the archive scan behavior has only been included in pre-release, where behavior is subject to change. 

-----

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2622)

### CODE REVIEW

**Comment by denelon** [comment]
How about --ignore-archive-malware-scan ?

I'd also like to know why we're thinking about enabling this behavior. Do we think there are false positives? Or is this just to be "more" complete? I believe we would want another administrator required setting to be able to "enable" a user to be able to pass this argument as well.

We're adding a policy for this specific scenario related to the certificate pinning for the "msstore" source. We would likely need a policy for this scenario as well.

Mentioning @AmelBawa-msft since he's looking at the other policy.

**Comment by denelon** [comment]
Or maybe "bypass" as the other policy is written.

**Comment by Trenly** [comment]
> I'd also like to know why we're thinking about enabling this behavior. Do we think there are false positives? Or is this just to be "more" complete? 

Using the latest dev build on the latest release of paint.net (local manifest) flags it as malware even though it is not, likely due to the inclusion of the dll's in the zip file

> I believe we would want another administrator required setting to be able to "enable" a user to be able to pass this argument as well.

Understandable. When `--force` was changed to no longer be the hash override, though, it no longer requires an admin setting. I can certainly see value in adding it, but I would imagine it more as a separate feature

**Comment by Trenly** [comment]
> How about --ignore-archive-malware-scan ?

I suppose this gets back to user experience. Would there be a necessity for having individual arguments for all security checks? I'm imagining a case where a user is attempting to install a .zip package with a hash mismatch, fails malware scan (false positive), and has known CVE. The implementation in this PR would allow `winget install <package> --ignore-security-hash --ignore-security-checks`. Using an individual argument would necessitate `winget install <package> --ignore-security-hash --ignore-archive-scan --ignore-cves`. I think either way would be fine, but I was presuming that if a user was willing to bypass an archive scan or CVE data, they would want to bypass the other security checks as well

> Or maybe "bypass" as the other policy is written.

I wrote this with the same concept as hash override. If `bypass` is preferred, I would suggest all security related arguments use `bypass`. This can still be done, as 1.4 is not in stable yet
https://github.com/microsoft/winget-cli/blob/edfc884fd3a54ff1cda372bbe7588d8025d0e1e1/src/AppInstallerCLICore/Argument.cpp#L60-L61

**Comment by yao-msft** [comment]
I would say for this one, it's not actually "bypassing", we still perform the check but "ignore" the results.

**Comment by JohnMcPMS** [comment]
> > I'd also like to know why we're thinking about enabling this behavior. Do we think there are false positives? Or is this just to be "more" complete?
> 
> Using the latest dev build on the latest release of paint.net (local manifest) flags it as malware even though it is not, likely due to the inclusion of the dll's in the zip file

This seems like a problem; @ryfu-msft should look at that.  From my understanding, Pure has a fairly extensive set of "bad" it will detect.  We should filter that to the appropriate ones for our scenario.

>Understandable. When --force was changed to no longer be the hash override, though, it no longer requires an admin setting. I can certainly see value in adding it, but I would imagine it more as a separate feature

And @yao-msft should look into that if true.  The admin setting should still be required.

> How about --ignore-archive-malware-scan ?

I prefer a specific flag per security related thing.  Yes, it means more flags to pass, but it also means that you opted in to ignoring each one failing, rather than opting in to the ones that existed when you wrote your script.

**Comment by yao-msft** [comment]
> > Understandable. When --force was changed to no longer be the hash override, though, it no longer requires an admin setting. I can certainly see value in adding it, but I would imagine it more as a separate feature
> 
> And @yao-msft should look into that if true. The admin setting should still be required.

I think there might be some misunderstanding here? The HashOverride was never behind an admin setting, the only admin settings we have are EnableLocalManifests and EnableBypassStoreCertPinning. HashOverride was behind a group policy, and it still is behind it after the change. And we block HashOverride if it's running as admin.

Adding group policy for ignore-malware-scan is new work though.

**Comment by Trenly** [comment]
> I prefer a specific flag per security related thing. Yes, it means more flags to pass, but it also means that you opted in to ignoring each one failing, rather than opting in to the ones that existed when you wrote your script.

I switched to the more specific `--ignore-malware-scan`

> Adding group policy for ignore-malware-scan is new work though.

I believe I added everything needed for group policy

File: doc/admx/en-US/DesktopAppInstaller.adml
Comment (yao-msft) [inline doc/admx/en-US/DesktopAppInstaller.adml:44]:
```suggestion
        If you enable or do not configure this policy, users will be able to enable the ability to override the Archive Malware Scan security validation using command line arguments.
        
        If you disable this policy, users will not be able to enable the ability to override the Archive Malware Scan security validation using command line arguments.

```

nit: Just to be consistent with OverrideHashMismatch above, and it's not actually "bypassing" the step

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw
Comment (yao-msft) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:892]:
```suggestion
    <value>Enable Windows App Installer Malware Scan Override</value>

```

File: src/Microsoft.Management.Deployment/PackageManager.idl
Comment (yao-msft) [inline src/Microsoft.Management.Deployment/PackageManager.idl:732]: Please revert all the changes under Microsoft.Management.Deployment

First, this only applies to installation using -m <manifest>, this code will never be invoked from Com entrypoint. It is only applicable to local manifest testing. Second, even if this code will be invoked, this is bypassing the group policy.

File: src/PowerShell/Microsoft.WinGet.Client/Common/BaseInstallCommand.cs
Comment (yao-msft) [inline src/PowerShell/Microsoft.WinGet.Client/Common/BaseInstallCommand.cs:63]: Revert changes in this file as well

File: doc/admx/en-US/DesktopAppInstaller.adml
Comment (Trenly) [reply doc/admx/en-US/DesktopAppInstaller.adml]: Committed

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw
Comment (Trenly) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw]: Committed

File: src/Microsoft.Management.Deployment/PackageManager.idl
Comment (Trenly) [reply src/Microsoft.Management.Deployment/PackageManager.idl]: Reverted

File: src/PowerShell/Microsoft.WinGet.Client/Common/BaseInstallCommand.cs
Comment (Trenly) [reply src/PowerShell/Microsoft.WinGet.Client/Common/BaseInstallCommand.cs]: Reverted

File: doc/admx/en-US/DesktopAppInstaller.adml
Comment (yao-msft) [inline doc/admx/en-US/DesktopAppInstaller.adml:49]: This is probably my bad, I'm too lazy to put a suggestion for each line ...

**Comment by yao-msft** [comment]
@Trenly as fyi, all group policy values and strings will need to be reviewed by some Microsoft internal team, so these naming/description may need to be changed after review (we'll put comments to the pr).

Also group policy changes require OS side changes, we'll  merge this pr after OS side changes are merged.

**Comment by Trenly** [comment]
> @Trenly as fyi, all group policy values and strings will need to be reviewed by some Microsoft internal team, so these naming/description may need to be changed after review (we'll put comments to the pr).
> 
> Also group policy changes require OS side changes, we'll merge this pr after OS side changes are merged.

Thank you for the info!

File: src/AppInstallerCLICore/Argument.cpp
Comment (yao-msft) [inline src/AppInstallerCLICore/Argument.cpp:73]:
```suggestion
            return Argument{ "ignore-local-archive-malware-scan"_liv, NoAlias, Args::Type::IgnoreLocalArchiveMalwareScan, Resource::String::IgnoreLocalArchiveMalwareScanArgumentDescription, ArgumentType::Flag, Settings::TogglePolicy::Policy::IgnoreLocalArchiveMalwareScan};

```

File: src/AppInstallerCLICore/Commands/InstallCommand.cpp
Comment (yao-msft) [inline src/AppInstallerCLICore/Commands/InstallCommand.cpp:44]:
```suggestion
            Argument::ForType(Args::Type::IgnoreLocalArchiveMalwareScan),

```

File: src/AppInstallerCLICore/Workflows/ArchiveFlow.cpp
Comment (yao-msft) [inline src/AppInstallerCLICore/Workflows/ArchiveFlow.cpp:32]:
```suggestion
                    AICLI_LOG(CLI, Warning, << "Archive malware scan failed; proceeding due to --ignore-local-archive-malware-scan override");
```

File: doc/specs/#140 - ZIP Support.md
Comment (yao-msft) [inline doc/specs/#140 - ZIP Support.md:71]:
```suggestion
If the Pure library functions that are called on a given ZIP file detect malware, the process will terminate and a warning will be displayed to the user. The user can include `--ignore-local-archive-malware-scan` to bypass this check and continue installation at their own risk.

```

File: src/AppInstallerCLIE2ETests/GroupPolicy.cs
Comment (yao-msft) [inline src/AppInstallerCLIE2ETests/GroupPolicy.cs:78]:
```suggestion
            var result = TestCommon.RunAICLICommand("install", "AnyPackage --ignore-local-archive-malware-scan");

```

**Comment by yao-msft** [comment]
Hi @Trenly 

The group policy review has finished. The review team has made some naming suggestions. Mainly we should be more specific about what this policy does.

Can you apply the changes and merge with latest code base? If you don't have time, may I push the changes to your branch?

Thanks

**Comment by Trenly** [comment]
> Hi @Trenly
> 
> The group policy review has finished. The review team has made some naming suggestions. Mainly we should be more specific about what this policy does.
> 
> Can you apply the changes and merge with latest code base? If you don't have time, may I push the changes to your branch?
> 
> Thanks

I think I applied them. There were several merge conflicts with the latest code base. I think I fixed everything now, but I wouldn't be surprised if I made a mistake. If something fails, just give me a nudge (or feel free to commit right to the branch, whichever is easiest for you)

**Comment by Trenly** [comment]
Same failure? I'm a bit confused
```xml

    <testcase classname="AppInstallerCLITests.exe.global" name="InstallFlow_Zip_ArchiveScanOverride" time="0.0">
      <failure message="installOutput.str().find(Resource::LocString(Resource::String::ArchiveFailedMalwareScanOverridden).get()) != std::string::npos" type="REQUIRE">
FAILED:
  REQUIRE( installOutput.str().find(Resource::LocString(Resource::String::ArchiveFailedMalwareScanOverridden).get()) != std::string::npos )
with expansion:
  18446744073709551615 (0xffffffffffffffff)
  !=
  18446744073709551615 (0xffffffffffffffff)
\x1B[0mFound \x1B[96mAppInstaller Test Zip Installer\x1B[0m [\x1B[96mAppInstallerCliTest.TestZipInstaller\x1B[0m] Version 1.0.0.0
\x1B[0mThis application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
\x1B[0mSuccessfully verified installer hash
\x1B[91mArchive scan detected malware. To override this check use --ignore-local-archive-malware-scan

at D:\a\1\s\src\AppInstallerCLITests\WorkFlow.cpp(1311)
      </failure>
      <failure type="FAIL_CHECK">
FAILED:
Unused override 
\x1B[0mFound \x1B[96mAppInstaller Test Zip Installer\x1B[0m [\x1B[96mAppInstallerCliTest.TestZipInstaller\x1B[0m] Version 1.0.0.0
\x1B[0mThis application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
\x1B[0mSuccessfully verified installer hash
\x1B[91mArchive scan detected malware. To override this check use --ignore-local-archive-malware-scan

at D:\a\1\s\src\AppInstallerCLITests\WorkFlow.cpp(504)
      </failure>
      <failure type="FAIL_CHECK">
FAILED:
Unused override 
\x1B[0mFound \x1B[96mAppInstaller Test Zip Installer\x1B[0m [\x1B[96mAppInstallerCliTest.TestZipInstaller\x1B[0m] Version 1.0.0.0
\x1B[0mThis application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
\x1B[0mSuccessfully verified installer hash
\x1B[91mArchive scan detected malware. To override this check use --ignore-local-archive-malware-scan

at D:\a\1\s\src\AppInstallerCLITests\WorkFlow.cpp(504)
      </failure>
    </testcase>
```

**Comment by Trenly** [comment]
> /azp run

Same error; I don't quite understand it 😞 

```xml
    <testcase classname="AppInstallerCLITests.exe.global" name="InstallFlow_Zip_ArchiveScanOverride" time="0.0">
      <failure message="installOutput.str().find(Resource::LocString(Resource::String::ArchiveFailedMalwareScanOverridden).get()) != std::string::npos" type="REQUIRE">
FAILED:
  REQUIRE( installOutput.str().find(Resource::LocString(Resource::String::ArchiveFailedMalwareScanOverridden).get()) != std::string::npos )
with expansion:
  4294967295 (0xffffffff)
  !=
  4294967295 (0xffffffff)
\x1B[0mFound \x1B[96mAppInstaller Test Zip Installer\x1B[0m [\x1B[96mAppInstallerCliTest.TestZipInstaller\x1B[0m] Version 1.0.0.0
\x1B[0mThis application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
\x1B[0mSuccessfully verified installer hash
\x1B[91mArchive scan detected malware. To override this check use --ignore-local-archive-malware-scan

at D:\a\1\s\src\AppInstallerCLITests\WorkFlow.cpp(1311)
      </failure>
      <failure type="FAIL_CHECK">
FAILED:
Unused override 
\x1B[0mFound \x1B[96mAppInstaller Test Zip Installer\x1B[0m [\x1B[96mAppInstallerCliTest.TestZipInstaller\x1B[0m] Version 1.0.0.0
\x1B[0mThis application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
\x1B[0mSuccessfully verified installer hash
\x1B[91mArchive scan detected malware. To override this check use --ignore-local-archive-malware-scan

at D:\a\1\s\src\AppInstallerCLITests\WorkFlow.cpp(504)
      </failure>
      <failure type="FAIL_CHECK">
FAILED:
Unused override 
\x1B[0mFound \x1B[96mAppInstaller Test Zip Installer\x1B[0m [\x1B[96mAppInstallerCliTest.TestZipInstaller\x1B[0m] Version 1.0.0.0
\x1B[0mThis application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
\x1B[0mSuccessfully verified installer hash
\x1B[91mArchive scan detected malware. To override this check use --ignore-local-archive-malware-scan

at D:\a\1\s\src\AppInstallerCLITests\WorkFlow.cpp(504)
      </failure>
    </testcase>
```

**Comment by yao-msft** [comment]
The parameter here needs to be updated with the new arg https://github.com/microsoft/winget-cli/blob/050056a6ceb46dcd62127da112255dceed8114f1/src/AppInstallerCLITests/WorkFlow.cpp#L1301

**Comment by Trenly** [comment]
> The parameter here needs to be updated with the new arg
> 
> https://github.com/microsoft/winget-cli/blob/050056a6ceb46dcd62127da112255dceed8114f1/src/AppInstallerCLITests/WorkFlow.cpp#L1301

I must be blind, though I do see you got to it before I did; Thank you

**Review by yao-msft — APPROVED** [review]

**Pull request merged**
